core/trino-main/src/main/java/io/trino/operator/output/columnar/PositionsAppenderFactory.java

```diff
@@ -0,0 +1,297 @@
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    block.writeBytesTo(position, 0, block.getSliceLength(position), blockBuilder);
+                    blockBuilder.closeEntry();
+                }
+            }
+        }
+    }
+
+    public static class SmallintPositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    blockBuilder.writeShort(block.getShort(position, 0)).closeEntry();
+                }
+            }
+        }
+    }
+
+    public static class Int96PositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    blockBuilder.writeLong(block.getLong(position, 0));
+                    blockBuilder.writeInt(block.getInt(position, SIZE_OF_LONG));
+                    blockBuilder.closeEntry();
+                }
+            }
+        }
+    }
+
+    public static class Int128PositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    blockBuilder.writeLong(block.getLong(position, 0));
+                    blockBuilder.writeLong(block.getLong(position, SIZE_OF_LONG));
+                    blockBuilder.closeEntry();
+                }
+            }
+        }
+    }
+
+    private static class CacheKey
+    {
+        private final Type type;
+        private final Class<? extends Block> blockClass;
```

Comment:
That should be `Class<? extends Block>`. You should extract block types from all children using `io.trino.spi.block.Block#getChildren`

Reply:
Do you mean List< Class<? extends Block>>? This version does not support complex types like `RowType` or `MapType` so this won't help.
For `RowType` support, this would probably be some kind of tree unless we don't care about the lower level nesting.